File: backend/src/services/SyncService.js
import { query, transaction } from '../database/db.js';
import { v4 as uuidv4 } from 'uuid';
import { hashPassword } from '../utils/auth.js';

class SyncService {
  /**
   * Pull delta updates from central MySQL database
   */
  async pullUpdates(sinceTimestamp = null) {
    const defaultTime = '1970-01-01 00:00:00';
    const cutoff = sinceTimestamp || defaultTime;

    // Fetch modified properties
    const properties = await query(
      `SELECT p.*, 
              AVG(r.rating) as avg_rating,
              COUNT(DISTINCT r.id) as review_count,
              GROUP_CONCAT(pi.image_url SEPARATOR '||') as images
       FROM properties p
       LEFT JOIN reviews r ON p.id = r.property_id AND r.status = 'approved'
       LEFT JOIN property_images pi ON p.id = pi.property_id
       WHERE p.updated_at >= ?
       GROUP BY p.id`,
      [cutoff]
    );

    // Fetch modified bookings
    const bookings = await query(
      `SELECT b.*, p.name as property_name, u.first_name, u.last_name, u.email, u.phone
       FROM bookings b
       JOIN properties p ON b.property_id = p.id
       JOIN users u ON b.customer_id = u.id
       WHERE b.updated_at >= ?`,
      [cutoff]
    );

    // Fetch modified payments
    const payments = await query(
      `SELECT p.*, b.total_amount, pr.name as property_name, u.first_name, u.last_name, u.email
       FROM payments p
       JOIN bookings b ON p.booking_id = b.id
       JOIN properties pr ON b.property_id = pr.id
       JOIN users u ON b.customer_id = u.id
       WHERE p.updated_at >= ?`,
      [cutoff]
    );

    // Fetch modified shifts
    const shifts = await query(
      `SELECT s.*, u.first_name as cashier_first_name, u.last_name as cashier_last_name
       FROM shifts s
       JOIN users u ON s.cashier_id = u.id
       WHERE s.updated_at >= ?`,
      [cutoff]
    );

    const nowResult = await query('SELECT NOW() as sync_server_time');
    const serverTime = nowResult[0]?.sync_server_time || new Date().toISOString();

    return {
      serverTime,
      delta: {
        properties,
        bookings,
        payments,
        shifts,
      },
    };
  }

  /**
   * Push queued offline records to MySQL
   */
  async pushUpdates(deviceId, payload) {
    const { properties = [], bookings = [], payments = [], shifts = [], logs = [] } = payload;
    let syncedCount = 0;
    const conflicts = [];

    await transaction(async (connection) => {
      // 1. Process Properties
      for (const prop of properties) {
        const [existing] = await connection.execute('SELECT id, updated_at FROM properties WHERE id = ?', [prop.id]);
        if (existing.length === 0) {
          await connection.execute(
            `INSERT INTO properties (
              id, owner_id, name, type, description, address, city, country,
              latitude, longitude, price_per_night, capacity, status, created_at, updated_at
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, NOW(), NOW())`,
            [
              prop.id || uuidv4(),
              prop.ownerId || prop.owner_id,
              prop.name,
              prop.type,
              prop.description || '',
              prop.address,
              prop.city || 'Nairobi',
              prop.country || 'Kenya',
              prop.latitude || null,
              prop.longitude || null,
              prop.pricePerNight || prop.price_per_night,
              prop.capacity || 1,
              prop.status || 'available',
            ]
          );
        } else {
          await connection.execute(
            `UPDATE properties SET 
              name = ?, type = ?, description = ?, address = ?, city = ?, country = ?,
              latitude = ?, longitude = ?, price_per_night = ?, capacity = ?, status = ?, updated_at = NOW()
             WHERE id = ?`,
            [
              prop.name,
              prop.type,
              prop.description || '',
              prop.address,
              prop.city || 'Nairobi',
              prop.country || 'Kenya',
              prop.latitude || null,
              prop.longitude || null,
              prop.pricePerNight || prop.price_per_night,
              prop.capacity || 1,
              prop.status || 'available',
              prop.id,
            ]
          );
        }
        syncedCount++;
      }

      // 2. Process Bookings
      for (const b of bookings) {
        const customerId = await this.resolveBookingCustomer(connection, b);
        const [existing] = await connection.execute('SELECT id, updated_at, status FROM bookings WHERE id = ?', [b.id]);
        if (existing.length === 0) {
          await connection.execute(
            `INSERT INTO bookings (
              id, property_id, customer_id, check_in_date, check_out_date,
              number_of_guests, total_amount, status, source, special_requests,
              pickup_location, pickup_time, pickup_notes, created_at, updated_at
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, NOW(), NOW())`,
            [
              b.id || uuidv4(),
              b.propertyId || b.property_id,
              customerId,
              b.checkInDate || b.check_in_date,
              b.checkOutDate || b.check_out_date,
              b.numberOfGuests || b.number_of_guests || 1,
              b.totalAmount || b.total_amount,
              b.status || 'pending',
              b.source || 'admin',
              b.specialRequests || b.special_requests || null,
              b.pickupLocation || b.pickup_location || null,
              b.pickupTime || b.pickup_time || null,
              b.pickupNotes || b.pickup_notes || null,
            ]
          );
        } else {
          await connection.execute(
            `UPDATE bookings SET 
              status = ?, check_in_date = ?, check_out_date = ?, 
              number_of_guests = ?, total_amount = ?, special_requests = ?,
              pickup_location = ?, pickup_time = ?, pickup_notes = ?, updated_at = NOW()
             WHERE id = ?`,
            [
              b.status,
              b.checkInDate || b.check_in_date,
              b.checkOutDate || b.check_out_date,
              b.numberOfGuests || b.number_of_guests || 1,
              b.totalAmount || b.total_amount,
              b.specialRequests || b.special_requests || null,
              b.pickupLocation || b.pickup_location || null,
              b.pickupTime || b.pickup_time || null,
              b.pickupNotes || b.pickup_notes || null,
              b.id,
            ]
          );
        }
        syncedCount++;
      }

      // 3. Process Payments
      for (const p of payments) {
        const [existing] = await connection.execute('SELECT id FROM payments WHERE id = ?', [p.id]);
        if (existing.length === 0) {
          await connection.execute(
            `INSERT INTO payments (
              id, booking_id, amount, method, status, transaction_ref, issued_by, notes, created_at, updated_at
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, NOW(), NOW())`,
            [
              p.id || uuidv4(),
              p.bookingId || p.booking_id,
              p.amount,
              p.method,
              p.status || 'paid',
              p.transactionRef || p.transaction_ref || null,
              p.issuedBy || p.issued_by || null,
              p.notes || null,
            ]
          );
        } else {
          await connection.execute(
            `UPDATE payments SET amount = ?, method = ?, status = ?, transaction_ref = ?, notes = ?, updated_at = NOW()
             WHERE id = ?`,
            [
              p.amount,
              p.method,
              p.status,
              p.transactionRef || p.transaction_ref || null,
              p.notes || null,
              p.id,
            ]
          );
        }
        syncedCount++;
      }

      // 4. Process Shifts
      for (const s of shifts) {
        const [existing] = await connection.execute('SELECT id FROM shifts WHERE id = ?', [s.id]);
        if (existing.length === 0) {
          await connection.execute(
            `INSERT INTO shifts (
              id, cashier_id, open_time, close_time, opening_balance,
              expected_closing_balance, actual_closing_balance, transaction_count, status, notes, created_at, updated_at
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, NOW(), NOW())`,
            [
              s.id || uuidv4(),
              s.cashierId || s.cashier_id,
              s.openTime || s.open_time || new Date(),
              s.closeTime || s.close_time || null,
              s.openingBalance || s.opening_balance || 0,
              s.expectedClosingBalance || s.expected_closing_balance || null,
              s.actualClosingBalance || s.actual_closing_balance || null,
              s.transactionCount || s.transaction_count || 0,
              s.status || 'open',
              s.notes || null,
            ]
          );
        } else {
          await connection.execute(
            `UPDATE shifts SET 
              close_time = ?, expected_closing_balance = ?, actual_closing_balance = ?,
              transaction_count = ?, status = ?, notes = ?, updated_at = NOW()
             WHERE id = ?`,
            [
              s.closeTime || s.close_time || null,
              s.expectedClosingBalance || s.expected_closing_balance || null,
              s.actualClosingBalance || s.actual_closing_balance || null,
              s.transactionCount || s.transaction_count || 0,
              s.status,
              s.notes || null,
              s.id,
            ]
          );
        }
        syncedCount++;
      }

      // 5. Ingest Audit / Sync Log
      await connection.execute(
        `INSERT INTO sync_logs (id, entity_type, entity_id, action, status, device_id, timestamp)
         VALUES (?, 'batch_sync', ?, 'push', 'synced', ?, NOW())`,
        [uuidv4(), deviceId || 'unknown_device', deviceId || 'client']
      );
    });

    return {
      success: true,
      syncedCount,
      conflicts,
      timestamp: new Date().toISOString(),
    };
  }

  async resolveBookingCustomer(connection, booking) {
    const requestedId = booking.customerId || booking.customer_id;
    if (requestedId) {
      const [existing] = await connection.execute('SELECT id FROM users WHERE id = ?', [requestedId]);
      if (existing.length > 0) return requestedId;
    }

    const phone = booking.customerPhone || booking.customer_phone || null;
    if (phone) {
      const [existing] = await connection.execute('SELECT id FROM users WHERE phone = ?', [phone]);
      if (existing.length > 0) return existing[0].id;
    }

    const customerId = uuidv4();
    const name = (booking.customerName || booking.customer_name || 'POS Guest').trim().split(/\s+/);
    const email = booking.customerEmail || booking.customer_email || `pos-${customerId}@rms.local`;
    await connection.execute(
      `INSERT INTO users (id, email, phone, first_name, last_name, role, status, password_hash, created_at, updated_at)
       VALUES (?, ?, ?, ?, ?, 'customer', 'active', ?, NOW(), NOW())`,
      [customerId, email, phone, name[0] || 'POS', name.slice(1).join(' ') || 'Guest', await hashPassword(uuidv4())]
    );
    return customerId;
  }

  /**
   * Get sync activity logs
   */
  async getSyncLogs(page = 1, limit = 50) {
    const offset = (page - 1) * limit;
    const [count] = await query('SELECT COUNT(*) as total FROM sync_logs');
    const logs = await query(
      'SELECT * FROM sync_logs ORDER BY timestamp DESC LIMIT ? OFFSET ?',
      [limit, offset]
    );

    return {
      data: logs,
      pagination: {
        page,
        limit,
        total: count.total,
        pages: Math.ceil(count.total / limit),
      },
    };
  }
}

export default new SyncService();

File: pos-system/POSSystem/Sync/SyncService.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Threading.Tasks;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using RMS.POSSystem.Database;
using RMS.POSSystem.Services;

namespace RMS.POSSystem.Sync
{
    public class SyncService
    {
        private readonly POSDatabase _db;
        private readonly HttpClient _client;
        private readonly string _apiBaseUrl;

        public SyncService(POSDatabase db, string apiBaseUrl = "http://localhost:3000/api")
        {
            _db = db;
            _apiBaseUrl = apiBaseUrl;
            _client = new HttpClient { Timeout = TimeSpan.FromSeconds(15) };
        }

        public async Task<bool> SyncNowAsync()
        {
            try
            {
                if (string.IsNullOrWhiteSpace(AuthService.CurrentToken)) return false;
                _client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", AuthService.CurrentToken);

                var state = _db.ExecuteQuery("SELECT LastPullAt FROM SyncState WHERE Id = 'central'");
                var since = state.Count > 0 ? state[0]["LastPullAt"]?.ToString() : "1970-01-01 00:00:00";
                var pullResponse = await _client.PostAsync($"{_apiBaseUrl}/sync/pull", JsonContent(new { sinceTimestamp = since }));
                if (!pullResponse.IsSuccessStatusCode) return false;

                var data = JObject.Parse(await pullResponse.Content.ReadAsStringAsync())["data"];
                ImportProperties(data?["delta"]?["properties"] as JArray);
                ImportBookings(data?["delta"]?["bookings"] as JArray, data?["delta"]?["payments"] as JArray);

                var serverTime = data?["serverTime"]?.ToString();
                if (!string.IsNullOrWhiteSpace(serverTime))
                {
                    _db.ExecuteNonQuery(
                        "INSERT INTO SyncState (Id, LastPullAt) VALUES ('central', @time) ON CONFLICT(Id) DO UPDATE SET LastPullAt=@time",
                        new Dictionary<string, object> { { "time", serverTime } });
                }

                return await PushPendingTransactionsAsync();
            }
            catch
            {
                return false;
            }
        }

        private void ImportProperties(JArray properties)
        {
            if (properties == null) return;
            foreach (JObject property in properties)
            {
                _db.ExecuteNonQuery(
                    @"INSERT INTO Properties (Id, Name, Type, Address, PricePerNight, Capacity, Status, CreatedAt, UpdatedAt)
                      VALUES (@id, @name, @type, @address, @price, @capacity, @status, @now, @now)
                      ON CONFLICT(Id) DO UPDATE SET Name=@name, Type=@type, Address=@address,
                      PricePerNight=@price, Capacity=@capacity, Status=@status, UpdatedAt=@now",
                    new Dictionary<string, object>
                    {
                        { "id", property["id"]?.ToString() }, { "name", property["name"]?.ToString() },
                        { "type", property["type"]?.ToString() }, { "address", property["address"]?.ToString() },
                        { "price", property["price_per_night"]?.Value<decimal>() ?? 0m },
                        { "capacity", property["capacity"]?.Value<int>() ?? 1 },
                        { "status", property["status"]?.ToString() ?? "available" }, { "now", DateTime.UtcNow }
                    });
            }
        }

        private void ImportBookings(JArray bookings, JArray payments)
        {
            if (bookings == null) return;
            var paymentsByBooking = new Dictionary<string, JObject>();
            if (payments != null)
            {
                foreach (JObject payment in payments)
                {
                    var bookingId = payment["booking_id"]?.ToString();
                    if (!string.IsNullOrWhiteSpace(bookingId)) paymentsByBooking[bookingId] = payment;
                }
            }

            foreach (JObject booking in bookings)
            {
                var bookingId = booking["id"]?.ToString();
                if (string.IsNullOrWhiteSpace(bookingId)) continue;
                paymentsByBooking.TryGetValue(bookingId, out var payment);
                var customerName = $"{booking["first_name"]} {booking["last_name"]}".Trim();
                _db.ExecuteNonQuery(
                    @"INSERT INTO Transactions
                      (Id, PropertyId, CustomerId, CashierId, CheckInDate, CheckOutDate, NumberOfGuests,
                       TotalAmount, AmountPaid, PaymentMethod, TransactionReference, TransactionStatus,
                       SyncStatus, ReceiptNumber, CreatedAt, UpdatedAt, SpecialRequests,
                       PickupLocation, PickupDestination, PickupTime, PickupNotes)
                      VALUES (@id, @property, @customer, @cashier, @checkin, @checkout, @guests, @total, @paid,
                       @method, @reference, @status, 'synced', @receipt, @created, @updated, @notes,
                       @pickupLocation, @pickupDestination, @pickupTime, @pickupNotes)
                      ON CONFLICT(Id) DO UPDATE SET PropertyId=@property, CustomerId=@customer,
                       CheckInDate=@checkin, CheckOutDate=@checkout, NumberOfGuests=@guests, TotalAmount=@total,
                       AmountPaid=@paid, PaymentMethod=@method, TransactionReference=@reference,
                       TransactionStatus=@status, SyncStatus='synced', UpdatedAt=@updated,
                       SpecialRequests=@notes, PickupLocation=@pickupLocation,
                       PickupDestination=@pickupDestination, PickupTime=@pickupTime,
                       PickupNotes=@pickupNotes",
                    new Dictionary<string, object>
                    {
                        { "id", bookingId }, { "property", booking["property_id"]?.ToString() },
                        { "customer", customerName }, { "cashier", AuthService.CurrentUser?.Id },
                        { "checkin", booking["check_in_date"]?.ToString() }, { "checkout", booking["check_out_date"]?.ToString() },
                        { "guests", booking["number_of_guests"]?.Value<int>() ?? 1 },
                        { "total", booking["total_amount"]?.Value<decimal>() ?? 0m },
                        { "paid", payment?["amount"]?.Value<decimal>() ?? 0m },
                        { "method", payment?["method"]?.ToString() ?? "online" },
                        { "reference", payment?["transaction_ref"]?.ToString() ?? "" },
                        { "status", booking["status"]?.ToString() ?? "pending" }, { "receipt", $"WEB-{bookingId}" },
                        { "created", booking["created_at"] != null ? (object)booking["created_at"].ToString() : DateTime.UtcNow },
                        { "updated", booking["updated_at"] != null ? (object)booking["updated_at"].ToString() : DateTime.UtcNow },
                        { "notes", booking["special_requests"]?.ToString() ?? "" },
                        { "pickupLocation", booking["pickup_location"]?.ToString() ?? "" },
                        { "pickupDestination", booking["pickup_destination"]?.ToString() ?? "" },
                        { "pickupTime", booking["pickup_time"]?.ToString() ?? "" },
                        { "pickupNotes", booking["pickup_notes"]?.ToString() ?? "" }
                    });
            }
        }

        private async Task<bool> PushPendingTransactionsAsync()
        {
            var pending = _db.ExecuteQuery("SELECT * FROM Transactions WHERE SyncStatus = 'pending' LIMIT 50");
            if (pending.Count == 0) return true;

            var bookings = new List<object>();
            var payments = new List<object>();
            foreach (var tx in pending)
            {
                bookings.Add(new
                {
                    id = tx["Id"], propertyId = tx["PropertyId"], customerId = tx["CustomerId"],
                    customerName = tx["CustomerId"],
                    checkInDate = tx["CheckInDate"], checkOutDate = tx["CheckOutDate"],
                    numberOfGuests = tx["NumberOfGuests"], totalAmount = tx["TotalAmount"],
                    status = tx["TransactionStatus"], source = "pos", specialRequests = tx["SpecialRequests"],
                    pickupLocation = tx["PickupLocation"], pickupDestination = tx["PickupDestination"],
                    pickupTime = tx["PickupTime"], pickupNotes = tx["PickupNotes"]
                });
                if (Convert.ToDecimal(tx["AmountPaid"] ?? 0m) > 0)
                {
                    payments.Add(new
                    {
                        id = $"{tx["Id"]}-payment", bookingId = tx["Id"], amount = tx["AmountPaid"],
                        method = tx["PaymentMethod"], status = "paid", transactionRef = tx["TransactionReference"],
                        issuedBy = tx["CashierId"]
                    });
                }
            }

            var response = await _client.PostAsync(
                $"{_apiBaseUrl}/sync/push",
                JsonContent(new { deviceId = "pos-terminal-01", payload = new { bookings, payments } }));
            if (!response.IsSuccessStatusCode) return false;

            _db.ExecuteNonQuery("UPDATE Transactions SET SyncStatus = 'synced' WHERE SyncStatus = 'pending'");
            _db.ExecuteNonQuery("UPDATE SyncLogs SET Status = 'synced' WHERE Status = 'pending'");
            return true;
        }

        private static StringContent JsonContent(object value) =>
            new StringContent(JsonConvert.SerializeObject(value), Encoding.UTF8, "application/json");
    }
}
